v2/googlecloud-to-splunk/src/main/java/com/google/cloud/teleport/v2/templates/GCSToSplunk.java

```diff
@@ -0,0 +1,291 @@
+
+    PCollection<SplunkWriteError> wrappedSplunkWriteErrors =
+        writeToSplunk(splunkEventTuple.get(SPLUNK_EVENT_OUT), options);
+
+    flattenErrorsAndConvertToString(
+            failsafeTransformedLines.get(UDF_ERROR_OUT),
+            splunkEventTuple.get(SPLUNK_EVENT_ERROR_OUT),
+            wrappedSplunkWriteErrors)
+        .apply("Output Errors To GCS", writeErrorsToGCS(options));
+
+    return pipeline.run();
+  }
+
+  static ReadCsv readFromCsv(GCSToSplunkOptions options) {
+    return CsvConverters.ReadCsv.newBuilder()
+        .setCsvFormat(options.getCsvFormat())
+        .setDelimiter(options.getDelimiter())
+        .setHasHeaders(options.getContainsHeaders())
+        .setInputFileSpec(options.getInputFileSpec())
+        .setHeaderTag(CSV_HEADERS)
+        .setLineTag(CSV_LINES)
+        .setFileEncoding(options.getCsvFileEncoding())
+        .build();
+  }
+
+  static PCollectionTuple convertToFailsafeAndMaybeApplyUdf(
+      PCollectionTuple csvLines, GCSToSplunkOptions options) {
+
+    String transformStepName = "Convert To Failsafe Element";
+
+    // In order to avoid generating a graph that makes it look like a UDF was called when none was
+    // intended, simply convert the input to FailsafeElements.
+    if (Strings.isNullOrEmpty(options.getJavascriptTextTransformGcsPath())) {
+      if (!Strings.isNullOrEmpty(options.getJsonSchemaPath())) {
+        transformStepName = "Apply JSON Schema";
+      }
+      return csvLines.apply(
+          transformStepName,
+          CsvConverters.LineToFailsafeJson.newBuilder()
+              .setDelimiter(options.getDelimiter())
+              .setJsonSchemaPath(options.getJsonSchemaPath())
+              .setHeaderTag(CSV_HEADERS)
+              .setLineTag(CSV_LINES)
+              .setUdfOutputTag(UDF_OUT)
+              .setUdfDeadletterTag(UDF_ERROR_OUT)
+              .build());
+    }
+
+    // For testing purposes, we need to do this check before creating the PTransform. Otherwise, we
+    // get a NullPointerException due to the PTransform not returning
+    // a value.
+    if (Strings.isNullOrEmpty(options.getJavascriptTextTransformFunctionName())) {
+      throw new IllegalArgumentException(
+          "JavaScript function name cannot be null or empty if file is set");
+    }
+    transformStepName = "Apply UDF";
+
+    return csvLines.apply(
+        transformStepName,
+        CsvConverters.LineToFailsafeJson.newBuilder()
+            .setDelimiter(options.getDelimiter())
+            .setUdfFileSystemPath(options.getJavascriptTextTransformGcsPath())
+            .setUdfFunctionName(options.getJavascriptTextTransformFunctionName())
+            .setJsonSchemaPath(options.getJsonSchemaPath())
+            .setHeaderTag(CSV_HEADERS)
+            .setLineTag(CSV_LINES)
+            .setUdfOutputTag(UDF_OUT)
+            .setUdfDeadletterTag(UDF_ERROR_OUT)
+            .build());
+  }
+
+  static PCollectionTuple convertToSplunkEvent(
```

Comment:
I think it makes sense, if there's only one `apply`, to return the `PTransform` rather than a `PCollection` or `PCollectionTuple`. That'll make `run` a bit cleaner and potentially make pipeline construction more fluent. We also really only care about testing the `PTransform`, not the label, so nothing is lost in testing.

This also goes for `writeToSplunk`